### PR TITLE
Create assert_almost_equal function for distributions

### PR DESCRIPTION
**Pull request by josephmure**
This functionality is used to remove prints from the GammaFactory_std C++ and Python tests, following the idea I mentioned in https://github.com/openturns/openturns/issues/2179#issuecomment-1301725198.

Other commits are unrelated doc fixes.

### CODE REVIEW

File: python/src/ConstantBasisFactory_doc.i.in [hunk 1]
```diff
@@ -17,7 +17,7 @@ A factory for constant basis of input dimension *dimension*.
 Examples
 --------
 >>> import openturns as ot
->>> basis = ConstantBasisFactory(2).build()
+>>> basis = ot.ConstantBasisFactory(2).build()
```
Comment (josephmure) [inline python/src/ConstantBasisFactory_doc.i.in:20]: @jschueller commented on this pull request:
> do you have an idea why this was not throwing an error ?

No, I agree it is strange. Either this code is not really run or something like ``from openturns import ConstantBasisFactory`` / ``from openturns import *`` was somehow run before.

Comment (jschueller) [reply python/src/ConstantBasisFactory_doc.i.in:20]: maybe its allowed when the class is from the same module, nevermind

File: lib/include/openturns/OTtestcode.hxx [hunk 2]
```diff
@@ -379,6 +380,17 @@ inline void assert_almost_equal(const SymmetricMatrix &a, const SymmetricMatrix
   }
 }
 
+inline void assert_almost_equal(const Distribution &a, const Distribution &b, const Scalar rtol = 1.0e-5, const Scalar atol = 1.0e-8, const String errMsg = "")
+{
+  if (a.getImplementation()->getClassName() != b.getImplementation()->getClassName())
```
Comment (alicelebrun) [inline lib/include/openturns/OTtestcode.hxx:385]: Wrong. A ComposedDistribution with Normal marginals and Normal copula may be equal to a Normal distribution

File: lib/include/openturns/OTtestcode.hxx [hunk 3]
```diff
@@ -379,6 +380,17 @@ inline void assert_almost_equal(const SymmetricMatrix &a, const SymmetricMatrix
...
+    throw InvalidArgumentException(HERE) << "A and B must be the same distribution. A is a " << a.getImplementation()->getClassName() << " whereas B is a " << b.getImplementation()->getClassName();
+
+  if (a.getParameterDimension() != b.getParameterDimension())
```
Comment (alicelebrun) [inline lib/include/openturns/OTtestcode.hxx:388]: Wrong. A Beta distribution  (4 parameters) may be equal to a Uniform distribution (2 parameters)

File: lib/include/openturns/OTtestcode.hxx [hunk 4]
```diff
@@ -379,6 +380,17 @@ inline void assert_almost_equal(const SymmetricMatrix &a, const SymmetricMatrix
...
+    throw InvalidArgumentException(HERE) << "A and B must have the same number of parameters. A has " << a.getParameterDimension() << " parameters whereas B has " << b.getParameterDimension() << " parameters.";
+
+  assert_almost_equal(a.getParameter(), b.getParameter(), rtol, atol, errMsg);
+}
```
Comment (alicelebrun) [inline lib/include/openturns/OTtestcode.hxx:392]: Wrong. The WeibullMax and GeneralizedExtremeValue distributions have the same number of parameters but are different if the parameters are equal

**Review by alicelebrun — CHANGES_REQUESTED** [review]
I suggest to remove the assert_almost_equal assertion for distribution as it will probably never be correct...

**Comment by jschueller** [comment]
alice seems to know a lot about openturns these days :)

**Comment by josephmure** [comment]
How about renaming the function `assert_has_almost_equal_parameters`? That way it becomes clear we are not really evaluating whether 2 distributions are equal in a mathematical sense.

**Comment by josephmure** [comment]
It could also be improved to recursively apply itself to the marginals and to the copula of a `ComposedDistribution`. Alternatively, we could get rid of the type check, which is not really the point of the function.

File: lib/include/openturns/OTtestcode.hxx (hunk 2 above)
Comment (josephmure) [reply lib/include/openturns/OTtestcode.hxx:385]: True, but is it a problem for unit tests?

File: lib/include/openturns/OTtestcode.hxx (hunk 3 above)
Comment (josephmure) [reply lib/include/openturns/OTtestcode.hxx:388]: True, but is it a problem for unit tests? If you want, I could change the name of the function to make clear it is not about equality in a mathematical sense, but in an object sense.

File: lib/include/openturns/OTtestcode.hxx (hunk 4 above)
Comment (josephmure) [reply lib/include/openturns/OTtestcode.hxx:392]: The assert would throw in this case because of the class name check above.

**Comment by regislebrun** [comment]
IMO it would actually be much easier to discuss it during a face to face meeting...

**Comment by mbaudin47** [comment]
A very different idea is suggested here, where we check the internal consistency of a Distribution: 

https://openturns.discourse.group/t/automated-consistency-tests-of-a-distribution/82

An idea which would fit to @regislebrun 's requirements would be to check the consistency between two Distributions:
- check the range
- check the moments
- check the PDF, the CDF, 
- etc.

This will never prove that two distributions are equal, but that some properties of them are.

**Review by regislebrun — APPROVED** [review]
Ok as it is not the approximate counterpart of the Distribution::equal method

**Review by jschueller — APPROVED** [review]

**Comment by jschueller** [comment]
could you rebase to run the new linter ?

**Review by sofianehaddad — APPROVED** [review]

**Pull request merged**